source/extensions/filters/network/zookeeper_proxy/zookeeper_decoder.cc

```diff
@@ -0,0 +1,360 @@
+#include "extensions/filters/network/zookeeper_proxy/zookeeper_decoder.h"
+
+#include "extensions/filters/network/zookeeper_proxy/zookeeper_utils.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ZooKeeperProxy {
+
+void DecoderImpl::decode(Buffer::Instance& data, uint64_t& offset) {
+  ENVOY_LOG(trace, "zookeeper_proxy: decoding {} bytes at offset {}", data.length(), offset);
+
+  // Check message length.
+  int32_t len;
+  BufferHelper::peekInt32(data, offset, len);
+  if (len < 8) {
+    callbacks_.onDecodeError();
+    return;
+  }
+
+  int32_t xid;
+  BufferHelper::peekInt32(data, offset, xid);
+
+  // "Special" requests.
+  switch (xid) {
+  case enumToIntSigned(XidCodes::CONNECT_XID):
+    parseConnect(data, offset, len);
+    return;
+  case enumToIntSigned(XidCodes::PING_XID):
+    callbacks_.onPing();
+    return;
+  case enumToIntSigned(XidCodes::AUTH_XID):
+    parseAuthRequest(data, offset, len);
+    return;
+  default:
+    break;
+  }
+
+  // "Regular" requests.
+  int32_t opcode;
+  BufferHelper::peekInt32(data, offset, opcode);
+  switch (opcode) {
+  case enumToInt(OpCodes::GETDATA):
+    parseGetDataRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::CREATE):
+    parseCreateRequest(data, offset, len, false);
+    break;
+  case enumToInt(OpCodes::CREATE2):
+    parseCreateRequest(data, offset, len, true);
+    break;
+  case enumToInt(OpCodes::SETDATA):
+    parseSetRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::GETCHILDREN):
+    parseGetChildrenRequest(data, offset, len, false);
+    break;
+  case enumToInt(OpCodes::GETCHILDREN2):
+    parseGetChildrenRequest(data, offset, len, true);
+    break;
+  case enumToInt(OpCodes::DELETE):
+    parseDeleteRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::EXISTS):
+    parseExistsRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::GETACL):
+    parseGetAclRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::SETACL):
+    parseSetAclRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::SYNC):
+    parseSyncRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::CHECK):
+    parseCheckRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::MULTI):
+    parseMultiRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::RECONFIG):
+    parseReconfigRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::SETWATCHES):
+    parseSetWatchesRequest(data, offset, len);
+    break;
+  case enumToIntSigned(OpCodes::CLOSE):
+    callbacks_.onCloseRequest();
+    break;
+  default:
+    break;
+  }
+}
+
+#define CHECK_LENGTH(LEN, MINL)                                                                    \
```

Comment:
Since you're throwing you can use a regular helper function for this instead of a macro.